File: samples/C# Samples/ScanAndUpload/ScanAndUpload/Form1.cs
using Dynamsoft.TWAIN;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Dynamsoft.TWAIN.Interface;
using Dynamsoft.Core;
using Dynamsoft.PDF;
using Dynamsoft.Core.Enums;
using System.IO;
using System.Runtime.InteropServices;



namespace ScanAndUpload
{
    public partial class Form1 : Form,IAcquireCallback,ISave
    {
        private TwainManager m_TwainManager = null;
        private ImageCore m_ImageCore = null;
        string m_strServerName = "www.dynamsoft.com";
        string m_strPort = "80";
        string m_strActionPage = "/demo/DNT/SaveToDB.aspx";
        private string m_StrProductKey = "t0068UwAAAEQABDxqjGfgEzhVYureL0kGxugcsvIqCDGTPTsR5nLaQsNupIc17Y5vpMZAWBDsd6Xw3NMYzdHlHwiKUrfe/cU=";

        public Form1()
        {
            InitializeComponent();
            m_TwainManager = new TwainManager(m_StrProductKey);
            m_ImageCore = new ImageCore();
            dsViewer1.Bind(m_ImageCore);
            this.txtboxServer.Text = m_strServerName;
            this.txtboxPort.Text = m_strPort;
            this.txtboxActionPage.Text = m_strActionPage;
            this.txtboxFileName.Text = "DNT_Image";
            this.rdbtnJPEG.Checked = true;
            this.chkboxMultiPage.Enabled = false;

            int lngNum;
            m_TwainManager.OpenSourceManager();
            for (lngNum = 0; lngNum < m_TwainManager.SourceCount; lngNum++)
            {
                cmbSource.Items.Add(m_TwainManager.SourceNameItems(Convert.ToInt16(lngNum))); // display the available imaging devices
            }
            if (lngNum > 0)
                cmbSource.SelectedIndex = 0;
        }


        private void btnScan_Click(object sender, EventArgs e)
        {
            m_ImageCore.ImageBuffer.IfAppendImage = true;
            AcquireImage();
        }

        private void AcquireImage()
        {
            try
            {
                m_TwainManager.SelectSourceByIndex(Convert.ToInt16(cmbSource.SelectedIndex));
                m_TwainManager.IfShowUI = chkboxIfShowUI.Checked;
                m_TwainManager.OpenSource();
                m_TwainManager.IfDisableSourceAfterAcquire = true;

                m_TwainManager.IfShowUI = chkboxIfShowUI.Checked;
                m_TwainManager.AcquireImage(this as IAcquireCallback);

            }
            catch (Exception exp)
            {
                this.txtboxErrMessage.AppendText(exp.Message);
                this.txtboxErrMessage.AppendText("\r\n");
            }

        }

        private void dynamicDotNetTwain1_OnMouseClick(short sImageIndex)
        {
            m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer = sImageIndex;
        }

        private void btnUpload_Click(object sender, EventArgs e)
        {
            try
            {
                string strImageName = "";
                string strHTTPServer = this.txtboxServer.Text; // the name or the IP of your HTTP Server
                if (strHTTPServer == "")
                {
                    this.txtboxErrMessage.AppendText("The HTTP server cannot be empty.\r\n");
                    return;
                }
                if (txtboxPort.Text == "")
                {
                    this.txtboxErrMessage.AppendText("The HTTP port cannot be empty.\r\n");
                    return;
                }
                else
                {
                    try
                    {
                        m_ImageCore.Net.HTTPPort = int.Parse(txtboxPort.Text); //the port number of the HTTP Server
                    }
                    catch
                    {
                        this.txtboxErrMessage.AppendText("Invalid HTTP port.\r\n");
                        return;
                    }
                }
                string strActionPage = this.txtboxActionPage.Text; // receive the uploaded images on the server side
                if (strActionPage == "")
                {
                    this.txtboxErrMessage.AppendText("The action page cannot be empty.\r\n");
                    return;
                }
                string strFileName = this.txtboxFileName.Text;
                if (strFileName == "")
                {
                    this.txtboxErrMessage.AppendText("The file name cannot be empty.\r\n");
                    return;
                }

                m_ImageCore.Net.HTTPUserName = this.txtboxName.Text; //user name for logging into HTTP Server
                m_ImageCore.Net.HTTPPassword = this.txtboxPwd.Text;
                m_ImageCore.Net.SetHTTPFormField("ExtraTxt", this.txtboxExtraTxt.Text); // pass extra text parameters when uploading image


                List<short> tempImageIndexList = new List<short>();
                tempImageIndexList.Add(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
                if (rdbtnJPEG.Checked)
                {
                    strImageName = strFileName + ".jpg";

                    m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer,strActionPage, strImageName,tempImageIndexList,Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_JPG);
                }
                if (rdbtnPNG.Checked)
                {
                    strImageName = strFileName + ".png";
                    m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName,tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_PNG);
                }
                if (rdbtnPDF.Checked)
                {
                    strImageName = strFileName + ".pdf";
                    PDFCreator tempCreator = new PDFCreator(m_StrProductKey);
                        Byte[] tempPDFBytes = tempCreator.SaveAsBytes(this as ISave);
                        m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName,tempPDFBytes);
                }
                if (rdbtnTIFF.Checked)
                {
                    strImageName = strFileName + ".tif";
                    if (chkboxMultiPage.Checked)
                    {
                        tempImageIndexList.Clear();
                        for(short i = 0;i<m_ImageCore.ImageBuffer.HowManyImagesInBuffer;i++)
                        {
                            tempImageIndexList.Add(i);
                        }
                        m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName,tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_TIF);
                    }
                    else
                    {
                        m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName, tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_TIF);
                    }
                }
                if (m_ImageCore.Net.HTTPPostResponseString == "")
                {
                    this.txtboxErrMessage.AppendText("Image uploaded to DB successfully. \r\n");
                    if (strHTTPServer.Trim().ToLower().CompareTo(m_strServerName.Trim().ToLower()) == 0 &&
                       txtboxPort.Text.Trim().ToLower().CompareTo(m_strPort.Trim().ToLower()) == 0 &&
                       strActionPage.Trim().ToLower().CompareTo(m_strActionPage.Trim().ToLower()) == 0)
                    {
                        SuccessInfo objSuccessInfo = new SuccessInfo(strImageName);
                        objSuccessInfo.ShowDialog();
                    }
                }
                else
                {
                    this.txtboxErrMessage.AppendText("Image uploaded to DB unsuccessfully. You can check DynamicDotNetTwain.HTTPPostResponseString for more information. \r\n");
                }
            }
            catch (Exception exp)
            {
                this.txtboxErrMessage.AppendText(exp.Message);
                this.txtboxErrMessage.AppendText("\r\n");
            }
        }

        private void rdbtnPDF_CheckedChanged(object sender, EventArgs e)
        {
            if (rdbtnPDF.Checked)
            {
                this.chkboxMultiPage.Enabled = true;
                this.chkboxMultiPage.Checked = true;
            }
            else
            {
                this.chkboxMultiPage.Enabled = false;
                this.chkboxMultiPage.Checked = false;
            }
        }

        private void rdbtnTIFF_CheckedChanged(object sender, EventArgs e)
        {
            if (rdbtnTIFF.Checked)
            {
                this.chkboxMultiPage.Enabled = true;
                this.chkboxMultiPage.Checked = true;
            }
            else
            {
                this.chkboxMultiPage.Enabled = false;
                this.chkboxMultiPage.Checked = false;
            }
        }

        private void txtboxServer_TextChanged(object sender, EventArgs e)
        {
            string strHTTPServer = this.txtboxServer.Text;
            if (strHTTPServer.Trim().ToLower().CompareTo(m_strServerName.Trim().ToLower()) == 0)
                lbNote.Visible = true;
            else
                lbNote.Visible = false;
        }

        public void OnPostAllTransfers()
        {
        }

        public bool OnPostTransfer(Bitmap bit)
        {
            m_ImageCore.IO.LoadImage(bit);
			TextBox.CheckForIllegalCrossThreadCalls = false;
            this.txtboxErrMessage.AppendText("Image acquired successfully. \r\n");
            return true;
        }

        public void OnPreAllTransfers()
        {
        }

        public bool OnPreTransfer()
        {
            return true;
        }

        public void OnSourceUIClose()
        {
        }

        public void OnTransferCancelled()
        {
        }

        public void OnTransferError()
        {
        }

        public object GetAnnotations(int iPageNumber)
        {
            if (chkboxMultiPage.Checked)
                return m_ImageCore.ImageBuffer.GetMetaData((short)iPageNumber, EnumMetaDataType.enumAnnotation);
            else
                return m_ImageCore.ImageBuffer.GetMetaData(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer,EnumMetaDataType.enumAnnotation);

        }

        public Bitmap GetImage(int iPageNumber)
        {
            if (chkboxMultiPage.Checked)
                return m_ImageCore.ImageBuffer.GetBitmap((short)iPageNumber);
            else
                return m_ImageCore.ImageBuffer.GetBitmap(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);

        }

        public int GetPageCount()
        {
            if (chkboxMultiPage.Checked)
                return m_ImageCore.ImageBuffer.HowManyImagesInBuffer;
            else
                return 1;
        }

        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            this.m_TwainManager.Dispose();
        }
    }
}

File: samples/VB .NET Samples/ScanAndUpload/ScanAndUpload/Form1.vb
Imports Dynamsoft.TWAIN
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports Dynamsoft.TWAIN.Interface
Imports Dynamsoft.Core
Imports Dynamsoft.PDF
Imports Dynamsoft.Core.Enums
Imports System.Runtime.InteropServices
Imports System.IO



Partial Public Class Form1
    Inherits Form
    Implements IAcquireCallback
    Implements ISave
    Private m_TwainManager As TwainManager = Nothing
    Private m_ImageCore As ImageCore = Nothing
    Private m_strServerName As String = "www.dynamsoft.com"
    Private m_strPort As String = "80"
    Private m_strActionPage As String = "/demo/DNT/SaveToDB.aspx"
    Private m_StrProductKey As String

    Public Sub New()
        InitializeComponent()
        m_StrProductKey = "t0068UwAAAEQABDxqjGfgEzhVYureL0kGxugcsvIqCDGTPTsR5nLaQsNupIc17Y5vpMZAWBDsd6Xw3NMYzdHlHwiKUrfe/cU="
        m_TwainManager = New TwainManager(m_StrProductKey)
        m_ImageCore = New ImageCore()
        dsViewer1.Bind(m_ImageCore)
        Me.txtboxServer.Text = m_strServerName
        Me.txtboxPort.Text = m_strPort
        Me.txtboxActionPage.Text = m_strActionPage
        Me.txtboxFileName.Text = "DNT_Image"
        Me.rdbtnJPEG.Checked = True
        Me.chkboxMultiPage.Enabled = False

        Dim lngNum As Integer
        m_TwainManager.OpenSourceManager()
        For lngNum = 0 To m_TwainManager.SourceCount - 1
            ' display the available imaging devices
            cmbSource.Items.Add(m_TwainManager.SourceNameItems(Convert.ToInt16(lngNum)))
        Next
        If lngNum > 0 Then
            cmbSource.SelectedIndex = 0
        End If
    End Sub



    Private Sub btnScan_Click(sender As Object, e As EventArgs)
        m_ImageCore.ImageBuffer.IfAppendImage = True
        AcquireImage()
    End Sub

    Private Sub AcquireImage()
        Try
            m_TwainManager.SelectSourceByIndex(Convert.ToInt16(cmbSource.SelectedIndex))
            m_TwainManager.IfShowUI = chkboxIfShowUI.Checked
            m_TwainManager.OpenSource()
            m_TwainManager.IfDisableSourceAfterAcquire = True

            m_TwainManager.IfShowUI = chkboxIfShowUI.Checked

            m_TwainManager.AcquireImage(TryCast(Me, IAcquireCallback))
        Catch exp As Exception
            Me.txtboxErrMessage.AppendText(exp.Message)
            Me.txtboxErrMessage.AppendText(vbCr & vbLf)
        End Try

    End Sub

    Private Sub dynamicDotNetTwain1_OnMouseClick(sImageIndex As Short)
        m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer = sImageIndex
    End Sub

    Private Sub btnUpload_Click(sender As Object, e As EventArgs)
        Try
            Dim strImageName As String = ""
            Dim strHTTPServer As String = Me.txtboxServer.Text
            ' the name or the IP of your HTTP Server
            If strHTTPServer = "" Then
                Me.txtboxErrMessage.AppendText("The HTTP server cannot be empty." & vbCr & vbLf)
                Return
            End If
            If txtboxPort.Text = "" Then
                Me.txtboxErrMessage.AppendText("The HTTP port cannot be empty." & vbCr & vbLf)
                Return
            Else
                Try
                    'the port number of the HTTP Server
                    m_ImageCore.Net.HTTPPort = Integer.Parse(txtboxPort.Text)
                Catch
                    Me.txtboxErrMessage.AppendText("Invalid HTTP port." & vbCr & vbLf)
                    Return
                End Try
            End If
            Dim strActionPage As String = Me.txtboxActionPage.Text
            ' receive the uploaded images on the server side
            If strActionPage = "" Then
                Me.txtboxErrMessage.AppendText("The action page cannot be empty." & vbCr & vbLf)
                Return
            End If
            Dim strFileName As String = Me.txtboxFileName.Text
            If strFileName = "" Then
                Me.txtboxErrMessage.AppendText("The file name cannot be empty." & vbCr & vbLf)
                Return
            End If

            m_ImageCore.Net.HTTPUserName = Me.txtboxName.Text
            'user name for logging into HTTP Server
            m_ImageCore.Net.HTTPPassword = Me.txtboxPwd.Text
            m_ImageCore.Net.SetHTTPFormField("ExtraTxt", Me.txtboxExtraTxt.Text)
            ' pass extra text parameters when uploading image

            Dim tempImageIndexList As New List(Of Short)()
            tempImageIndexList.Add(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
            If rdbtnJPEG.Checked Then
                strImageName = strFileName & ".jpg"

                m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName, tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_JPG)
            End If
            If rdbtnPNG.Checked Then
                strImageName = strFileName & ".png"
                m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName, tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_PNG)
            End If
            If rdbtnPDF.Checked Then
                strImageName = strFileName & ".pdf"
                Dim tempCreator As New PDFCreator(m_StrProductKey)
                Dim tempPDFBytes As [Byte]() = tempCreator.SaveAsBytes(TryCast(Me, ISave))
                m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName, tempPDFBytes)
            End If
            If rdbtnTIFF.Checked Then
                strImageName = strFileName & ".tif"
                If chkboxMultiPage.Checked Then
                    tempImageIndexList.Clear()
                    For i As Short = 0 To m_ImageCore.ImageBuffer.HowManyImagesInBuffer - 1
                        tempImageIndexList.Add(i)
                    Next
                    m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName, tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_TIF)
                Else
                    m_ImageCore.Net.HTTPUploadThroughPost(strHTTPServer, strActionPage, strImageName, tempImageIndexList, Dynamsoft.Core.Enums.EnumImageFileFormat.WEBTW_TIF)
                End If
            End If
            If m_ImageCore.Net.HTTPPostResponseString = "" Then
                Me.txtboxErrMessage.AppendText("Image uploaded to DB successfully. " & vbCr & vbLf)
                If strHTTPServer.Trim().ToLower().CompareTo(m_strServerName.Trim().ToLower()) = 0 AndAlso txtboxPort.Text.Trim().ToLower().CompareTo(m_strPort.Trim().ToLower()) = 0 AndAlso strActionPage.Trim().ToLower().CompareTo(m_strActionPage.Trim().ToLower()) = 0 Then
                    Dim objSuccessInfo As New SuccessInfo(strImageName)
                    objSuccessInfo.ShowDialog()
                End If
            Else
                Me.txtboxErrMessage.AppendText("Image uploaded to DB unsuccessfully. You can check DynamicDotNetTwain.HTTPPostResponseString for more information. " & vbCr & vbLf)
            End If
        Catch exp As Exception
            Me.txtboxErrMessage.AppendText(exp.Message)
            Me.txtboxErrMessage.AppendText(vbCr & vbLf)
        End Try
    End Sub

    Private Sub rdbtnPDF_CheckedChanged(sender As Object, e As EventArgs)
        If rdbtnPDF.Checked Then
            Me.chkboxMultiPage.Enabled = True
            Me.chkboxMultiPage.Checked = True
        Else
            Me.chkboxMultiPage.Enabled = False
            Me.chkboxMultiPage.Checked = False
        End If
    End Sub

    Private Sub rdbtnTIFF_CheckedChanged(sender As Object, e As EventArgs)
        If rdbtnTIFF.Checked Then
            Me.chkboxMultiPage.Enabled = True
            Me.chkboxMultiPage.Checked = True
        Else
            Me.chkboxMultiPage.Enabled = False
            Me.chkboxMultiPage.Checked = False
        End If
    End Sub

    Private Sub txtboxServer_TextChanged(sender As Object, e As EventArgs)
        Dim strHTTPServer As String = Me.txtboxServer.Text
        If strHTTPServer.Trim().ToLower().CompareTo(m_strServerName.Trim().ToLower()) = 0 Then
            lbNote.Visible = True
        Else
            lbNote.Visible = False
        End If
    End Sub

    Public Sub OnPostAllTransfers() Implements IAcquireCallback.OnPostAllTransfers
    End Sub

    Public Function OnPostTransfer(bit As Bitmap) As Boolean Implements IAcquireCallback.OnPostTransfer
        m_ImageCore.IO.LoadImage(bit)
		TextBox.CheckForIllegalCrossThreadCalls = False
        Me.txtboxErrMessage.AppendText("Image acquired successfully. " & vbCr & vbLf)
        Return True
    End Function

    Public Sub OnPreAllTransfers() Implements IAcquireCallback.OnPreAllTransfers
    End Sub

    Public Function OnPreTransfer() As Boolean Implements IAcquireCallback.OnPreTransfer
        Return True
    End Function

    Public Sub OnSourceUIClose() Implements IAcquireCallback.OnSourceUIClose
    End Sub

    Public Sub OnTransferCancelled() Implements IAcquireCallback.OnTransferCancelled
    End Sub

    Public Sub OnTransferError() Implements IAcquireCallback.OnTransferError
    End Sub

    Public Function GetAnnotations(iPageNumber As Integer) As Object Implements ISave.GetAnnotations
        If chkboxMultiPage.Checked Then
            Return m_ImageCore.ImageBuffer.GetMetaData(CShort(iPageNumber), EnumMetaDataType.enumAnnotation)
        Else
            Return m_ImageCore.ImageBuffer.GetMetaData(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer, EnumMetaDataType.enumAnnotation)
        End If

    End Function

    Public Function GetImage(iPageNumber As Integer) As Bitmap Implements ISave.GetImage
        If chkboxMultiPage.Checked Then
            Return m_ImageCore.ImageBuffer.GetBitmap(CShort(iPageNumber))
        Else
            Return m_ImageCore.ImageBuffer.GetBitmap(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
        End If

    End Function

    Public Function GetPageCount() As Integer Implements ISave.GetPageCount
        If chkboxMultiPage.Checked Then
            Return m_ImageCore.ImageBuffer.HowManyImagesInBuffer
        Else
            Return 1
        End If
    End Function

    Private Sub Form1_FormClosed(sender As Object, e As FormClosedEventArgs)
        Me.m_TwainManager.Dispose()
    End Sub
End Class
